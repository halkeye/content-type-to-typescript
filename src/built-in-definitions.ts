import { fromPairs, map } from 'lodash';
import { JSONSchema } from './types/json-schema';

export const Location: JSONSchema = {
  title: 'Location',
  properties: {
    lon: {
      type: 'number',
      description: '',
      additionalProperties: false,
      required: [],
    },
    lat: {
      type: 'number',
      description: '',
      additionalProperties: false,
      required: [],
    },
  },
  required: ['lat', 'lon'],
  additionalProperties: false,
};

export const AssetLink: JSONSchema = {
  title: 'AssetLink',
  properties: {
    type: {
      type: 'string',
      description: '',
      additionalProperties: false,
      required: [],
    },
    linkType: {
      type: 'string',
      description: '',
      additionalProperties: false,
      required: [],
    },
    id: {
      type: 'string',
      additionalProperties: false,
      required: [],
    },
  },
  required: ['type', 'linkType', 'id'],
  additionalProperties: false,
};

export const EntryLink: JSONSchema = {
  title: 'EntryLink',
  properties: {
    type: {
      type: 'string',
      description: '',
      additionalProperties: false,
      required: [],
    },
    linkType: {
      type: 'string',
      description: '',
      additionalProperties: false,
      required: [],
    },
    id: {
      type: 'string',
      additionalProperties: false,
      required: [],
    },
  },
  required: ['type', 'linkType', 'id'],
  additionalProperties: false,
};

export const buildRef = (schema: JSONSchema) => {
  return `#/definitions/${schema.title}`;
};

export const getByRef = (ref: string): JSONSchema => {
  const lookup = fromPairs(map([Location, EntryLink, AssetLink], (def) => [buildRef(def), def]));

  return lookup[ref];
};
