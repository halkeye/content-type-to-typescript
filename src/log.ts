import chalk from 'chalk';

// tslint:disable:no-console
export const log = (message: string) => console.log(message);
export const logError = (message: string) => console.error(chalk.redBright(message));
export const logSuccess = (message: string) => console.log(chalk.greenBright(message));
