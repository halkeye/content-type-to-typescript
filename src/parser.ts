import { compile } from 'json-schema-to-typescript';
import { chain, fromPairs } from 'lodash';
import { AssetLink, buildRef, EntryLink, Location } from './built-in-definitions';
import { ContentType, Field } from './types/contentful';
import { JSONSchema } from './types/json-schema';

function transformTitle(schema: JSONSchema, contentTypeInfo: ContentType) {
  return {
    ...schema,
    title: contentTypeInfo.name,
  };
}

function fieldToJsonSchema(fieldInfo: Field): any {
  let result: any;
  switch (fieldInfo.type) {
    case 'Symbol':
    case 'Text':
    case 'Date':
      result = {
        type: 'string',
      };
      break;
    case 'Number':
    case 'Integer':
      result = {
        type: 'number',
      };
      break;
    case 'Boolean':
      result = {
        type: 'boolean',
      };
      break;
    case 'Location':
      result = {
        $ref: buildRef(Location),
      };
      break;
    case 'Object':
      result = {
        type: 'object',
      };
      break;
    case 'Array':
      if (!fieldInfo.items) {
        throw new Error('Unexpected Content Type structure.');
      }
      result = {
        items: fieldToJsonSchema(fieldInfo.items),
        type: 'array',
      };
      break;
    case 'Link':
      if (fieldInfo.linkType === 'Asset') {
        result = {
          $ref: buildRef(AssetLink),
        };
      } else if (fieldInfo.linkType === 'Entry') {
        result = {
          $ref: buildRef(EntryLink),
        };
      } else {
        throw new Error('Unexpected Content Type structure.');
      }
      break;
    default:
      throw new Error(`Type ${fieldInfo.type} is not yet supported`);
  }

  return result;
}

function transformFields(contentTypeInfo: Partial<ContentType>): JSONSchema {
  const properties = chain(contentTypeInfo.fields)
    .filter((fieldInfo) => !fieldInfo.omitted)
    .map((fieldInfo) => [fieldInfo.id, fieldToJsonSchema(fieldInfo)])
    .fromPairs()
    .value();

  const required = chain(contentTypeInfo.fields)
    .filter((fieldInfo) => fieldInfo.required)
    .map((fieldInfo) => fieldInfo.id)
    .value();

  return {
    properties,
    required,
    additionalProperties: false,
  };
}

export function convertToJSONSchema(contentTypeInfo: Partial<ContentType>): JSONSchema {
  const resultSchema: JSONSchema = {
    title: contentTypeInfo.name,
    description: contentTypeInfo.description || contentTypeInfo.name,
    ...transformFields(contentTypeInfo),
  };

  return resultSchema;
}
