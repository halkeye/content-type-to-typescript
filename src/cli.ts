import * as program from 'commander';
import compileFromSpace from './compile-from-space';
import { logError } from './log';
// tslint:disable:no-console

const ensureArgNotEmpty = (value: string, message: string): void => {
  if (!value) {
    logError(message);
    process.exit(1);
  }
};

program
  .option(
    '--management-token <managementToken>',
    // tslint:disable-next-line:max-line-length
    'This is the management token for this space. You can generate the token in the Contentful web app. Learn more at https://www.contentful.com/developers/docs/references/authentication/',
  )
  .option('--space <space>', 'This is the space ID')
  .option('--environment <environment>', 'This is the environment, defaults to "master"')
  .option('--output <output>', 'Output filename: e.g. -o ./types.ts')
  .option('--prefix <prefix>', 'prefix used for interface names e.g. "Contentful"');

program.on('--help', () => {
  console.log('  Examples:');
  console.log('');
  console.log(
    '    $ content-type-to-typescript --management-token <token> --space <space> --environment <environment> --output <filename>',
  );
  console.log(
    '    $ content-type-to-typescript --management-token=<token> --space=<space> --environment <environment> --output=<filename>',
  );
  console.log('');
});

program.parse(process.argv);

// tslint:disable:no-console

ensureArgNotEmpty(program.managementToken, 'Management token is missing.');
ensureArgNotEmpty(program.space, 'Space is missing.');
ensureArgNotEmpty(program.output, 'Output file path is missing.');

compileFromSpace({
  managementToken: program.managementToken,
  spaceId: program.space,
  environmentId: program.environment || 'master',
  output: program.output,
  prefix: program.prefix,
})
  .then(() => process.exit())
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });
