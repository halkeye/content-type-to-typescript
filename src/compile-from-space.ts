import { ContentType } from 'contentful';
import { createClient } from 'contentful-management';
import * as fs from 'fs';
import { compileFromContentTypes } from './content-type-to-typescript';
import { logError, logSuccess } from './log';

async function fetchContentTypes({
  accessToken,
  spaceId,
  environmentId,
}: {
  accessToken: string;
  spaceId: string;
  environmentId: string;
}): Promise<ContentType[]> {
  try {
    const client = createClient({ accessToken });
    const space = await client.getSpace(spaceId);
    const environment = await space.getEnvironment(environmentId);

    let skip = 0;
    let contentTypes: any[] = [];
    while (true) {
      const { items, total } = await environment.getContentTypes({skip});
      contentTypes = contentTypes.concat(items);
      skip += items.length;

      if (skip >= total) {
        break;
      }
    }

    return contentTypes;
  } catch (err) {
    if (err.response && err.response.data) {
      logError(err.response.data.message);
    } else {
      logError(err);
    }

    throw err;
  }
}

async function compile(contentTypes: ContentType[], prefix?: string): Promise<string> {
  try {
    const ts = await compileFromContentTypes(contentTypes, undefined, prefix);
    return ts;
  } catch (err) {
    logError(err.message);
    throw err;
  }
}

function writeFile(output: string, ts: string): Promise<void> {
  return new Promise((resolve, reject) => {
    fs.writeFile(output, ts, (err) => {
      if (err) {
        logError(err.message);
        reject(err);
      } else {
        logSuccess('TypeScript Definitions were successfully created!');
        resolve();
      }
    });
  });
}

export default async function({
  accessToken,
  spaceId,
  environmentId,
  output,
  prefix,
}: {
  accessToken: string;
  spaceId: string;
  environmentId: string;
  output: string;
  prefix?: string
}) {
  const contentTypes = await fetchContentTypes({ accessToken, spaceId, environmentId });

  const ts = await compile(contentTypes, prefix);

  await writeFile(output, ts);
}
