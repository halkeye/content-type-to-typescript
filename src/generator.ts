import { ContentType, Field } from 'contentful/index';
import { compile, Options } from 'json-schema-to-typescript';
import { chain, defaults, fromPairs, get, orderBy } from 'lodash';
import { buildRef, getByRef } from './built-in-definitions';
import { convertToJSONSchema } from './parser';
import { JSONSchema } from './types/json-schema';

const BANNER_COMMENT = `/**
* This file was automatically generated.
* DO NOT MODIFY IT BY HAND.
*/`;

export async function compileFromContentTypes(
  contentTypes: Array<Partial<ContentType>>,
  options: Partial<Options> = {},
): Promise<string> {
  const settings = defaults(
    {
      bannerComment: BANNER_COMMENT,
    },
    options,
  );

  const allDefinitions = includeRequiredDefinitions(
    contentTypes.map((ct) => convertToJSONSchema(ct)),
  );

  const resultSchema = {
    anyOf: allDefinitions.map((def) => ({ $ref: buildRef(def) })),
    definitions: chain(allDefinitions)
      .map((def) => [def.title, def])
      .fromPairs()
      .value(),
  };

  const res = await compile(resultSchema, EPHEMERAL_ROOT, settings);

  return cleanupEphemeralRoot(res);
}

const EPHEMERAL_ROOT = 'EphemeralContentfulSchemaRoot1';

function cleanupEphemeralRoot(input: string): string {
  return input.replace(new RegExp(`\n.+${EPHEMERAL_ROOT}.+`, 'gm'), '');
}

function getRefs(definition: JSONSchema): JSONSchema[] {
  return chain(definition.properties)
    .map((propSchema) => get(propSchema, '$ref'))
    .compact()
    .map((s) => getByRef(s))
    .compact()
    .value();
}

function includeRequiredDefinitions(definitions: JSONSchema[]): JSONSchema[] {
  const requiredBuiltInDefinitions: JSONSchema[] = chain(definitions)
    .map((def) => getRefs(def))
    .flatten()
    .value();

  const alphabetizedDefinitions = orderBy(definitions, ['title'], ['asc']);

  return alphabetizedDefinitions.concat(requiredBuiltInDefinitions);
}
