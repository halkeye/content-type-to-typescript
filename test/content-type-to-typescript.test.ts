import { ContentType } from 'contentful';
import { compileFromContentTypes } from '../src/content-type-to-typescript';

import allFields from './resources/all-fields.json';
import omittedField from './resources/omitted-fields.json';

async function macro(message: string, contentType: ContentType): Promise<void> {
  test(message, async () => {
    const actual = await compileFromContentTypes([contentType]);
    expect(actual).toMatchSnapshot();
  });
}

macro('All fields', allFields);
macro('Content Type with omitted field', omittedField);
