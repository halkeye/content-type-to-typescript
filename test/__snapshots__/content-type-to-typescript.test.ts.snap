// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`100 content types 1`] = `
"import { Asset, Entry } from 'contentful';/**
 * This file was automatically generated.
 * DO NOT MODIFY IT BY HAND.
 */

/**
 * ContentType0a
 */
export interface ContentType0A {
  singleLine?: string;
}
/**
 * ContentType10a
 */
export interface ContentType10A {
  singleLine?: string;
}
/**
 * ContentType11a
 */
export interface ContentType11A {
  singleLine?: string;
}
/**
 * ContentType12a
 */
export interface ContentType12A {
  singleLine?: string;
}
/**
 * ContentType13a
 */
export interface ContentType13A {
  singleLine?: string;
}
/**
 * ContentType14a
 */
export interface ContentType14A {
  singleLine?: string;
}
/**
 * ContentType15a
 */
export interface ContentType15A {
  singleLine?: string;
}
/**
 * ContentType16a
 */
export interface ContentType16A {
  singleLine?: string;
}
/**
 * ContentType17a
 */
export interface ContentType17A {
  singleLine?: string;
}
/**
 * ContentType18a
 */
export interface ContentType18A {
  singleLine?: string;
}
/**
 * ContentType19a
 */
export interface ContentType19A {
  singleLine?: string;
}
/**
 * ContentType1a
 */
export interface ContentType1A {
  singleLine?: string;
}
/**
 * ContentType20a
 */
export interface ContentType20A {
  singleLine?: string;
}
/**
 * ContentType21a
 */
export interface ContentType21A {
  singleLine?: string;
}
/**
 * ContentType22a
 */
export interface ContentType22A {
  singleLine?: string;
}
/**
 * ContentType23a
 */
export interface ContentType23A {
  singleLine?: string;
}
/**
 * ContentType24a
 */
export interface ContentType24A {
  singleLine?: string;
}
/**
 * ContentType25a
 */
export interface ContentType25A {
  singleLine?: string;
}
/**
 * ContentType26a
 */
export interface ContentType26A {
  singleLine?: string;
}
/**
 * ContentType27a
 */
export interface ContentType27A {
  singleLine?: string;
}
/**
 * ContentType28a
 */
export interface ContentType28A {
  singleLine?: string;
}
/**
 * ContentType29a
 */
export interface ContentType29A {
  singleLine?: string;
}
/**
 * ContentType2a
 */
export interface ContentType2A {
  singleLine?: string;
}
/**
 * ContentType30a
 */
export interface ContentType30A {
  singleLine?: string;
}
/**
 * ContentType31a
 */
export interface ContentType31A {
  singleLine?: string;
}
/**
 * ContentType32a
 */
export interface ContentType32A {
  singleLine?: string;
}
/**
 * ContentType33a
 */
export interface ContentType33A {
  singleLine?: string;
}
/**
 * ContentType34a
 */
export interface ContentType34A {
  singleLine?: string;
}
/**
 * ContentType35a
 */
export interface ContentType35A {
  singleLine?: string;
}
/**
 * ContentType36a
 */
export interface ContentType36A {
  singleLine?: string;
}
/**
 * ContentType37a
 */
export interface ContentType37A {
  singleLine?: string;
}
/**
 * ContentType38a
 */
export interface ContentType38A {
  singleLine?: string;
}
/**
 * ContentType39a
 */
export interface ContentType39A {
  singleLine?: string;
}
/**
 * ContentType3a
 */
export interface ContentType3A {
  singleLine?: string;
}
/**
 * ContentType40a
 */
export interface ContentType40A {
  singleLine?: string;
}
/**
 * ContentType41a
 */
export interface ContentType41A {
  singleLine?: string;
}
/**
 * ContentType42a
 */
export interface ContentType42A {
  singleLine?: string;
}
/**
 * ContentType43a
 */
export interface ContentType43A {
  singleLine?: string;
}
/**
 * ContentType44a
 */
export interface ContentType44A {
  singleLine?: string;
}
/**
 * ContentType45a
 */
export interface ContentType45A {
  singleLine?: string;
}
/**
 * ContentType46a
 */
export interface ContentType46A {
  singleLine?: string;
}
/**
 * ContentType47a
 */
export interface ContentType47A {
  singleLine?: string;
}
/**
 * ContentType48a
 */
export interface ContentType48A {
  singleLine?: string;
}
/**
 * ContentType49a
 */
export interface ContentType49A {
  singleLine?: string;
}
/**
 * ContentType4a
 */
export interface ContentType4A {
  singleLine?: string;
}
/**
 * ContentType50a
 */
export interface ContentType50A {
  singleLine?: string;
}
/**
 * ContentType51a
 */
export interface ContentType51A {
  singleLine?: string;
}
/**
 * ContentType52a
 */
export interface ContentType52A {
  singleLine?: string;
}
/**
 * ContentType53a
 */
export interface ContentType53A {
  singleLine?: string;
}
/**
 * ContentType54a
 */
export interface ContentType54A {
  singleLine?: string;
}
/**
 * ContentType55a
 */
export interface ContentType55A {
  singleLine?: string;
}
/**
 * ContentType56a
 */
export interface ContentType56A {
  singleLine?: string;
}
/**
 * ContentType57a
 */
export interface ContentType57A {
  singleLine?: string;
}
/**
 * ContentType58a
 */
export interface ContentType58A {
  singleLine?: string;
}
/**
 * ContentType59a
 */
export interface ContentType59A {
  singleLine?: string;
}
/**
 * ContentType5a
 */
export interface ContentType5A {
  singleLine?: string;
}
/**
 * ContentType60a
 */
export interface ContentType60A {
  singleLine?: string;
}
/**
 * ContentType61a
 */
export interface ContentType61A {
  singleLine?: string;
}
/**
 * ContentType62a
 */
export interface ContentType62A {
  singleLine?: string;
}
/**
 * ContentType63a
 */
export interface ContentType63A {
  singleLine?: string;
}
/**
 * ContentType64a
 */
export interface ContentType64A {
  singleLine?: string;
}
/**
 * ContentType65a
 */
export interface ContentType65A {
  singleLine?: string;
}
/**
 * ContentType66a
 */
export interface ContentType66A {
  singleLine?: string;
}
/**
 * ContentType67a
 */
export interface ContentType67A {
  singleLine?: string;
}
/**
 * ContentType68a
 */
export interface ContentType68A {
  singleLine?: string;
}
/**
 * ContentType69a
 */
export interface ContentType69A {
  singleLine?: string;
}
/**
 * ContentType6a
 */
export interface ContentType6A {
  singleLine?: string;
}
/**
 * ContentType70a
 */
export interface ContentType70A {
  singleLine?: string;
}
/**
 * ContentType71a
 */
export interface ContentType71A {
  singleLine?: string;
}
/**
 * ContentType72a
 */
export interface ContentType72A {
  singleLine?: string;
}
/**
 * ContentType73a
 */
export interface ContentType73A {
  singleLine?: string;
}
/**
 * ContentType74a
 */
export interface ContentType74A {
  singleLine?: string;
}
/**
 * ContentType75a
 */
export interface ContentType75A {
  singleLine?: string;
}
/**
 * ContentType76a
 */
export interface ContentType76A {
  singleLine?: string;
}
/**
 * ContentType77a
 */
export interface ContentType77A {
  singleLine?: string;
}
/**
 * ContentType78a
 */
export interface ContentType78A {
  singleLine?: string;
}
/**
 * ContentType79a
 */
export interface ContentType79A {
  singleLine?: string;
}
/**
 * ContentType7a
 */
export interface ContentType7A {
  singleLine?: string;
}
/**
 * ContentType80a
 */
export interface ContentType80A {
  singleLine?: string;
}
/**
 * ContentType81a
 */
export interface ContentType81A {
  singleLine?: string;
}
/**
 * ContentType82a
 */
export interface ContentType82A {
  singleLine?: string;
}
/**
 * ContentType83a
 */
export interface ContentType83A {
  singleLine?: string;
}
/**
 * ContentType84a
 */
export interface ContentType84A {
  singleLine?: string;
}
/**
 * ContentType85a
 */
export interface ContentType85A {
  singleLine?: string;
}
/**
 * ContentType86a
 */
export interface ContentType86A {
  singleLine?: string;
}
/**
 * ContentType87a
 */
export interface ContentType87A {
  singleLine?: string;
}
/**
 * ContentType88a
 */
export interface ContentType88A {
  singleLine?: string;
}
/**
 * ContentType89a
 */
export interface ContentType89A {
  singleLine?: string;
}
/**
 * ContentType8a
 */
export interface ContentType8A {
  singleLine?: string;
}
/**
 * ContentType90a
 */
export interface ContentType90A {
  singleLine?: string;
}
/**
 * ContentType91a
 */
export interface ContentType91A {
  singleLine?: string;
}
/**
 * ContentType92a
 */
export interface ContentType92A {
  singleLine?: string;
}
/**
 * ContentType93a
 */
export interface ContentType93A {
  singleLine?: string;
}
/**
 * ContentType94a
 */
export interface ContentType94A {
  singleLine?: string;
}
/**
 * ContentType95a
 */
export interface ContentType95A {
  singleLine?: string;
}
/**
 * ContentType96a
 */
export interface ContentType96A {
  singleLine?: string;
}
/**
 * ContentType97a
 */
export interface ContentType97A {
  singleLine?: string;
}
/**
 * ContentType98a
 */
export interface ContentType98A {
  singleLine?: string;
}
/**
 * ContentType99a
 */
export interface ContentType99A {
  singleLine?: string;
}
/**
 * ContentType9a
 */
export interface ContentType9A {
  singleLine?: string;
}
"
`;

exports[`All fields 1`] = `
"import { Asset, Entry } from 'contentful';/**
 * This file was automatically generated.
 * DO NOT MODIFY IT BY HAND.
 */

export interface AllTheFieldsAndWidgets {
  singleLine?: string;
  multipleLine?: string;
  markdown?: string;
  radio?: string;
  dropdown?: string;
  symbol?: string;
  slug?: string;
  url?: string;
  youtube?: string;
  ooyala?: string;
  kaltura?: string;
  number?: number;
  rating?: number;
  decimal?: number;
  bool?: boolean;
  date?: string;
  location?: Location;
  entry?: Entry<any>;
  asset?: Asset;
  object?: {
    [k: string]: any;
  };
  entryCards?: Entry<any>[];
  entries?: Entry<any>[];
  assetGallery?: Asset[];
  assets?: Asset[];
  symbols?: string[];
}
export interface Location {
  lon: number;
  lat: number;
}
"
`;

exports[`Content Type with omitted field 1`] = `
"import { Asset, Entry } from 'contentful';/**
 * This file was automatically generated.
 * DO NOT MODIFY IT BY HAND.
 */

export interface ModelWithOmitteField {
  multipleLine?: string;
}
"
`;

exports[`Multiple types 1`] = `
"import { Asset, Entry } from 'contentful';/**
 * This file was automatically generated.
 * DO NOT MODIFY IT BY HAND.
 */

export interface Author {
  name?: string;
  twitterHandle?: string;
  profilePhoto?: Asset;
  biography?: string;
  createdEntries?: Entry<XYw5JsIecuGE68MmGMg20>[];
}
export interface Image {
  title: string;
  photo?: Asset;
  imageCaption?: string;
  imageCredits?: string;
}
export interface PhotoGallery {
  title: string;
  slug?: string;
  author: Entry<any>;
  coverImage?: Asset;
  description?: string;
  images?: Entry<XYw5JsIecuGE68MmGMg20>[];
  tags?: string[];
  date?: string;
  location?: Location;
}
export interface Location {
  lon: number;
  lat: number;
}
"
`;

exports[`Multiple types with non unique field id 1`] = `
"import { Asset, Entry } from 'contentful';/**
 * This file was automatically generated.
 * DO NOT MODIFY IT BY HAND.
 */

export interface Post {
  title?: string;
}
export interface Post2 {
  title?: string;
}
"
`;

exports[`Parse descirption 1`] = `
"import { Asset, Entry } from 'contentful';/**
 * This file was automatically generated.
 * DO NOT MODIFY IT BY HAND.
 */

/**
 * Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur
 */
export interface ModelWithOmittedField {
  multipleLine?: string;
}
"
`;
